result.go

```diff
@@ -0,0 +1,177 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package dig
+
+import (
+	"fmt"
+	"reflect"
+)
+
+type (
+	result interface {
+		Produces() map[key]struct{}
+	}
+
+	resultList struct {
+		ctype    reflect.Type
+		produces map[key]struct{}
+
+		Results []result
+	}
+
+	resultError struct{}
+
+	resultSingle struct {
+		Name string
+		Type reflect.Type
+	}
+
+	resultObject struct {
+		produces map[key]struct{}
+
+		Type   reflect.Type
+		Fields []resultObjectField
+	}
+
+	resultObjectField struct {
+		Name   string
+		Index  int
+		Result result
+	}
+)
+
+var (
+	_ result = resultSingle{}
+	_ result = resultError{}
+	_ result = resultObject{}
+	_ result = resultList{}
+)
+
+func newResultList(ctype reflect.Type) (resultList, error) {
+	rl := resultList{
+		ctype:    ctype,
+		Results:  make([]result, ctype.NumOut()),
+		produces: make(map[key]struct{}),
+	}
+
+	for i := 0; i < ctype.NumOut(); i++ {
+		r, err := newResult(ctype.Out(i))
+		if err != nil {
+			return rl, errWrapf(err, "bad result %d", i+1)
+		}
+		rl.Results[i] = r
+
+		for k := range r.Produces() {
+			if _, ok := rl.produces[k]; ok {
+				return rl, fmt.Errorf("returns multiple %v", k)
+			}
+			rl.produces[k] = struct{}{}
+		}
+	}
+
+	if len(rl.produces) == 0 {
+		return rl, fmt.Errorf("%v must provide at least one non-error type", ctype)
+	}
+
+	return rl, nil
+}
+
+func (rl resultList) Produces() map[key]struct{} { return rl.produces }
+
+func newResult(t reflect.Type) (result, error) {
+	switch {
+	case isError(t):
+		return resultError{}, nil
+	case IsOut(t):
+		return newResultObject(t)
+	case embedsType(t, _outPtrType):
+		return nil, fmt.Errorf(
+			"%v embeds *dig.Out which is not supported, embed dig.Out value instead", t)
+	case t.Kind() == reflect.Ptr && IsOut(t.Elem()):
+		return nil, fmt.Errorf("%v is a pointer to dig.Out, use value type instead", t)
+		// Make sure we're not producing dig.In's either.
+	case IsIn(t) || (t.Kind() == reflect.Ptr && IsIn(t.Elem())) || embedsType(t, _inPtrType):
+		return nil, fmt.Errorf("cannot provide parameter objects: %v embeds a dig.In", t)
+	default:
+		return resultSingle{Type: t}, nil
+	}
+}
+
+func (rs resultSingle) Produces() map[key]struct{} {
+	return map[key]struct{}{
+		{name: rs.Name, t: rs.Type}: {},
+	}
+}
+
+// resultError doesn't produce anything
+func (resultError) Produces() map[key]struct{} { return nil }
+
+func newResultObject(t reflect.Type) (resultObject, error) {
+	ro := resultObject{Type: t, produces: make(map[key]struct{})}
+
+	for i := 0; i < t.NumField(); i++ {
+		f := t.Field(i)
+		if f.Type == _outType {
+			// Skip over the dig.Out embed.
+			continue
+		}
+
+		if f.PkgPath != "" {
+			return ro, fmt.Errorf(
+				"private fields not allowed in dig.Out, did you mean to export %q (%v) from %v?",
+				f.Name, f.Type, t)
+		}
+
+		if isError(f.Type) {
+			return ro, fmt.Errorf(
+				"cannot provide errors from dig.Out: field %q (%v) of %v is an error field",
```

Comment:
It'd be nice to suggest an alternative here - the user probably (maybe?) should be writing a `func() (ResultStruct, error)` instead.

Exact wording up to you.

Reply:
Switched to,

    cannot return errors from dig.Out, return it from the constructor instead: